sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastHelper.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.execution
+
+import com.nvidia.spark.rapids.GpuColumnVector
+
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.joins.EmptyHashedRelation
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+object GpuBroadcastHelper {
+  /**
+   * Given a broadcast relation get a ColumnarBatch that can be used on the GPU.
+   *
+   * The broadcast relation may or may not contain any data, so we special case
+   * the empty relation case (hash or identity depending on the type of join).
+   *
+   * If a broadcast result is unexpected we throw, but at this moment other
+   * cases are not known, so this is a defensive measure.
+   *
+   * @param broadcastRelation - the broadcast as produced by a broadcast exchange
+   * @param broadcastPlan - the SparkPlan to use to obtain the schema for the broadcast
+   *                      batch
+   * @return a `ColumnarBatch` or throw if the broadcast can't be handled
+   */
+  def getBroadcastBatch(broadcastRelation: Broadcast[Any],
+                        broadcastPlan: SparkPlan): ColumnarBatch = {
+    val broadcastRelationValue = broadcastRelation.value
+    broadcastRelationValue match {
+      case broadcastBatch: SerializeConcatHostBuffersDeserializeBatch =>
+        val builtBatch = broadcastBatch.batch
+        GpuColumnVector.incRefCounts(builtBatch)
+        builtBatch
+      case EmptyHashedRelation =>
+        GpuColumnVector.emptyBatch(broadcastPlan.schema)
+      case identity: Array[Any] if identity.length == 0 =>
```

Comment:
I'm confused why we need this logic.  We support broadcast nested loop joins today yet don't have code that is explicitly checking for identity vs. hash broadcasts.  Everything is a SerializeConcatHostBufferDeserializeBatch which can also handle empty tables and row-count-only tables.  Do we not have control over the driver-side code that is setting up the broadcast or something?

Reply:
Because of this: https://github.com/NVIDIA/spark-rapids/pull/4256/files#diff-7ca82be5615724f3ba4376b82be1c8a34e3fb5df34a3dedfa9a21eb3e6e54b97R321. So we are making it not return a `SerializedConcatHostBufferDeserializeBatch` on purpose to allow AQE's optimization to work, but that means we need to handle the special cases later.